File: instabiz/instabiz/report/ib_replenishment_suggestions/ib_replenishment_suggestions.js
frappe.query_reports["IB Replenishment Suggestions"] = {
	filters: [
		{ fieldname: "item_group", label: __("Item Group"), fieldtype: "Link", options: "Item Group" },
		{ fieldname: "lookback_days", label: __("Look back (days)"), fieldtype: "Int", default: 90 },
		{ fieldname: "default_lead_days", label: __("Lead time if not set (days)"), fieldtype: "Int", default: 14 },
		{ fieldname: "review_days", label: __("Review cycle (days)"), fieldtype: "Int", default: 7 },
		{ fieldname: "show_all", label: __("Show items with enough stock"), fieldtype: "Check", default: 0 },
	],

	onload(report) {
		report.page.add_inner_button(__("Create Material Request"), () => {
			const dt = report.datatable;
			const checked = dt ? dt.rowmanager.getCheckedRows().map((i) => report.data[i]) : [];
			const rows = (checked.length ? checked : report.data || []).filter((r) => r && r.suggested > 0);
			if (!rows.length) {
				frappe.msgprint(__("Nothing to order. Tick rows that have a suggested quantity."));
				return;
			}
			frappe.model.with_doctype("Material Request", () => {
				const mr = frappe.model.get_new_doc("Material Request");
				mr.material_request_type = "Purchase";
				mr.schedule_date = frappe.datetime.add_days(frappe.datetime.get_today(), 7);
				rows.forEach((r) => {
					const row = frappe.model.add_child(mr, "items");
					row.item_code = r.item_code;
					row.item_name = r.item_name;
					row.qty = r.suggested;
					row.uom = r.uom;
					row.stock_uom = r.uom;
					row.conversion_factor = 1;
					row.schedule_date = mr.schedule_date;
				});
				frappe.set_route("Form", "Material Request", mr.name);
			});
		});
	},

	get_datatable_options(options) {
		return Object.assign(options, { checkboxColumn: true });
	},

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) return value;
		if (column.fieldname === "cover_days") {
			const cls =
				data.cover_days < data.lead_time ? "red" : data.cover_days < data.lead_time + 7 ? "orange" : "green";
			return `<span class="indicator-pill ${cls} no-indicator-dot">${value}</span>`;
		}
		if (column.fieldname === "trend") {
			const cls = { Rising: "blue", Falling: "gray", Steady: "green" }[data.trend] || "gray";
			return `<span class="indicator-pill ${cls} no-indicator-dot">${value}</span>`;
		}
		return value;
	},
};

File: instabiz/instabiz/report/ib_replenishment_suggestions/ib_replenishment_suggestions.py
"""IB Replenishment Suggestions — which stock items to reorder, and how much.

Demand is read from submitted Sales Orders (what customers have ordered),
weighting the last 30 days against the longer look-back so a rising or falling
SKU is caught early. An item is flagged when projected stock (on hand + on
order - reserved, all warehouses) falls below its reorder point:

	reorder point = daily demand x lead time + safety stock
	suggested qty = daily demand x (lead time + review days) + safety stock - projected

Suggestions only: nothing is ordered. The report's "Create Material Request"
button drafts one for the ticked rows, for a buyer to review.
"""
import math

import frappe
from frappe import _
from frappe.utils import add_days, cint, flt, nowdate

_EXCLUDED_GROUPS = ("PACKAGING",)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	data = _data(filters)
	return _columns(), data, None, None, _summary(data)


def _columns():
	return [
		{"label": _("Item"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 190},
		{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 200},
		{"label": _("UOM"), "fieldname": "uom", "fieldtype": "Data", "width": 70},
		{"label": _("Daily Demand"), "fieldname": "daily", "fieldtype": "Float", "precision": 2, "width": 110},
		{"label": _("Trend"), "fieldname": "trend", "fieldtype": "Data", "width": 85},
		{"label": _("Projected Stock"), "fieldname": "projected", "fieldtype": "Float", "precision": 1, "width": 120},
		{"label": _("Days of Cover"), "fieldname": "cover_days", "fieldtype": "Float", "precision": 1, "width": 110},
		{"label": _("Lead Time (d)"), "fieldname": "lead_time", "fieldtype": "Int", "width": 100},
		{"label": _("Reorder Point"), "fieldname": "reorder_point", "fieldtype": "Float", "precision": 1, "width": 110},
		{"label": _("Suggested Qty"), "fieldname": "suggested", "fieldtype": "Float", "precision": 0, "width": 115},
	]


def _data(filters):
	lookback = cint(filters.lookback_days) or 90
	from instabiz.overrides.ib_settings import get_int

	default_lead = cint(filters.default_lead_days) or get_int("replenishment_lead_days", 14)
	review = cint(filters.review_days) or get_int("replenishment_review_days", 7)
	today = nowdate()

	cond = ""
	vals = {"since": add_days(today, -lookback), "recent": add_days(today, -30), "ex": _EXCLUDED_GROUPS}
	if filters.item_group:
		cond += " AND i.item_group = %(item_group)s"
		vals["item_group"] = filters.item_group

	demand = frappe.db.sql(
		f"""
		SELECT soi.item_code,
			SUM(soi.stock_qty) AS qty_all,
			SUM(CASE WHEN so.transaction_date >= %(recent)s THEN soi.stock_qty ELSE 0 END) AS qty_30
		FROM `tabSales Order Item` soi
		INNER JOIN `tabSales Order` so ON so.name = soi.parent
		INNER JOIN `tabItem` i ON i.name = soi.item_code
		WHERE so.docstatus = 1 AND so.transaction_date >= %(since)s
		  AND i.is_stock_item = 1 AND i.disabled = 0 AND i.item_group NOT IN %(ex)s {cond}
		GROUP BY soi.item_code
		""",
		vals,
		as_dict=True,
	)
	if not demand:
		return []

	codes = [d.item_code for d in demand]
	items = {
		r.name: r
		for r in frappe.get_all(
			"Item",
			filters={"name": ["in", codes]},
			fields=["name", "item_name", "stock_uom", "lead_time_days", "safety_stock"],
		)
	}
	projected = {
		r.item_code: flt(r.projected)
		for r in frappe.db.sql(
			"SELECT item_code, SUM(projected_qty) AS projected FROM `tabBin`"
			" WHERE item_code IN %(c)s GROUP BY item_code",
			{"c": codes},
			as_dict=True,
		)
	}

	rows = []
	for d in demand:
		it = items.get(d.item_code)
		if not it:
			continue
		avg_all = flt(d.qty_all) / lookback
		avg_30 = flt(d.qty_30) / 30
		daily = 0.5 * avg_all + 0.5 * avg_30  # lean toward the recent month
		if daily <= 0:
			continue
		lead = cint(it.lead_time_days) or default_lead
		safety = flt(it.safety_stock)
		proj = projected.get(d.item_code, 0.0)
		reorder_point = daily * lead + safety
		if not filters.show_all and proj >= reorder_point:
			continue
		suggested = max(0.0, daily * (lead + review) + safety - proj)
		ratio = avg_30 / avg_all if avg_all else 0
		rows.append({
			"item_code": d.item_code,
			"item_name": it.item_name,
			"uom": it.stock_uom,
			"daily": daily,
			"trend": "Rising" if ratio > 1.2 else "Falling" if ratio < 0.8 else "Steady",
			"projected": proj,
			"cover_days": round(proj / daily, 1) if proj > 0 else 0,
			"lead_time": lead,
			"reorder_point": reorder_point,
			"suggested": math.ceil(suggested),
		})

	# most urgent first: least cover relative to how long a reorder takes
	rows.sort(key=lambda r: (r["cover_days"] - r["lead_time"], -r["suggested"]))
	return rows


def _summary(data):
	return [
		{"label": _("Items to reorder"), "value": len([r for r in data if r["suggested"] > 0]),
			"datatype": "Int", "indicator": "orange"},
		{"label": _("Run out before stock arrives"), "value": len([r for r in data if r["cover_days"] < r["lead_time"]]),
			"datatype": "Int", "indicator": "red"},
		{"label": _("Rising demand"), "value": len([r for r in data if r["trend"] == "Rising"]),
			"datatype": "Int", "indicator": "blue"},
	]
